File: src/main/java/de/saschahlusiak/freebloks/preferences/ThemePreference.java
package de.saschahlusiak.freebloks.preferences;

import android.app.AlertDialog.Builder;
import android.content.Context;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.preference.ListPreference;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;

import de.saschahlusiak.freebloks.view.scene.LegacyTheme;

@Deprecated // use the ones in types instead
public class ThemePreference extends ListPreference {

	class ThemeAdapter extends ArrayAdapter<CharSequence> {
		public ThemeAdapter(Context context) {
			super(context, android.R.layout.simple_list_item_single_choice);
			for (int i = 0; i < getEntries().length; i++) {
				this.add("");
			}
		}
		@Override
		public View getView(int position, View convertView, ViewGroup parent) {
			View v = super.getView(position, convertView, parent);
			LegacyTheme t = LegacyTheme.getLegacy(getContext(), getEntryValues()[position].toString());
			if (t != null)
				t.apply(v);
			return v;
		}
	}

	public ThemePreference(Context context, AttributeSet attrs) {
		super(context, attrs);
	}

	@Override
	protected void onPrepareDialogBuilder(final Builder builder) {
		super.onPrepareDialogBuilder(builder);
		builder.setAdapter(new ThemeAdapter(getContext()), new OnClickListener() {
			@Override
			public void onClick(final DialogInterface dialog, int which) {
				String value = getEntryValues()[which].toString();
				if (callChangeListener(value)) {
					setValue(value);
				}
				dialog.dismiss();
			}
		});
	}
}

File: src/main/java/de/saschahlusiak/freebloks/preferences/types/ThemePreference.kt
package de.saschahlusiak.freebloks.preferences.types

import android.content.Context
import android.util.AttributeSet
import androidx.preference.ListPreference
import de.saschahlusiak.freebloks.R
import de.saschahlusiak.freebloks.theme.ThemeManager

class ThemePreference(context: Context, attrs: AttributeSet?) : ListPreference(context, attrs) {
    init {
        val tm = ThemeManager.get(context)
        val themes = tm.getAllThemes()

        // TODO: support new themes
        /*
        val entries = arrayOfNulls<String>(themes.size)
        val values = arrayOfNulls<String>(themes.size)

        for (i in themes.indices) {
            val theme = themes[i]

            entries[i] = theme.getLabel(context)
            values[i] = theme.name
        }
    */

        this.entries = context.resources.getStringArray(R.array.theme_labels)
        this.entryValues = context.resources.getStringArray(R.array.theme_values)
    }
}
